pkg/trie/trie.go

```diff
@@ -1,1518 +1,81 @@
-func (t *Trie) handleTrackedDeltas(success bool, pendingDeltas tracking.Getter) {
-	if !success || t.generation == 0 {
-		// Do not persist tracked deleted node hashes if the operation failed or
-		// if the trie generation is zero (first block, no trie snapshot done yet).
-		return
-	}
-	t.deltas.MergeWith(pendingDeltas)
-}
-
-func (t *Trie) prepForMutation(currentNode *Node,
-	copySettings node.CopySettings,
-	pendingDeltas DeltaRecorder) (
-	newNode *Node, err error) {
-	if currentNode.Generation == t.generation {
-		// no need to track deleted node, deep copy the node and
-		// update the node generation.
-		newNode = currentNode
-	} else {
-		err = t.registerDeletedNodeHash(currentNode, pendingDeltas)
-		if err != nil {
-			return nil, fmt.Errorf("registering deleted node: %w", err)
-		}
-		newNode = currentNode.Copy(copySettings)
-		newNode.Generation = t.generation
-	}
-
-	newNode.SetDirty()
-	return newNode, nil
-}
-
-func (t *Trie) registerDeletedNodeHash(node *Node,
-	pendingDeltas DeltaRecorder) (err error) {
-	err = t.ensureMerkleValueIsCalculated(node)
-	if err != nil {
-		return fmt.Errorf("ensuring Merkle value is calculated: %w", err)
-	}
-
-	if len(node.MerkleValue) < 32 {
-		// Merkle values which are less than 32 bytes are inlined
-		// in the parent branch and are not stored on disk, so there
-		// is no need to track their deletion for the online pruning.
-		return nil
-	}
-
-	if !node.Dirty {
-		// Only register deleted nodes that were not previously modified
-		// since the last trie snapshot.
-		nodeHash := common.NewHash(node.MerkleValue)
-		pendingDeltas.RecordDeleted(nodeHash)
-	}
-
-	return nil
-}
-
-// DeepCopy deep copies the trie and returns
-// the copy. Note this method is meant to be used
-// in tests and should not be used in production
-// since it's rather inefficient compared to the copy
-// on write mechanism achieved through snapshots.
-func (t *Trie) DeepCopy() (trieCopy *Trie) {
-	if t == nil {
-		return nil
-	}
-
-	trieCopy = &Trie{
-		generation: t.generation,
-		db:         t.db,
-		version:    t.version,
-	}
-
-	if t.deltas != nil {
-		// Because DeepCopy() is only used in tests (in this and other packages),
-		// it's fine to type assert deltas to access its DeepCopy method.
-		trieCopy.deltas = t.deltas.DeepCopy()
-	}
-
-	if t.childTries != nil {
-		trieCopy.childTries = make(map[common.Hash]*Trie, len(t.childTries))
-		for hash, trie := range t.childTries {
-			trieCopy.childTries[hash] = trie.DeepCopy()
-		}
-	}
-
-	if t.root != nil {
-		copySettings := node.DeepCopySettings
-		trieCopy.root = t.root.Copy(copySettings)
-	}
-
-	return trieCopy
+type ChildTrieManager interface {
+	GetChild(keyToChild []byte) (Trie, error)
+	GetFromChild(keyToChild, key []byte) ([]byte, error)
+	PutIntoChild(keyToChild, key, value []byte) error
+	DeleteChild(keyToChild []byte) (err error)
+	ClearFromChild(keyToChild, key []byte) error
+	GetChildTries() map[common.Hash]Trie
```

Comment:
I would suggest splitting those methods into small cohesive groups like: `ChildTrieGetter`, `ChildTrieSetter` and `ChildTrieDeleter`